File: Android/AndroidFrontEndChallenge/app/src/main/java/com/ams/androidfrontendchallenge/Fibonacci.kt
package com.ams.androidfrontendchallenge

object Fibonacci {

    fun recursive(n: Long): Long {
        return when {
            n < 0 -> throwNegativeArgumentException()
            else  -> if (n < 2) n else recursive(n - 1) + recursive(n - 2)
        }
    }

    fun iterative(n: Long): Long {
        return when {
            n < 0    -> throwNegativeArgumentException()
            n <= 2 -> n
            else     -> {
                var minusOne = 1L
                var minusTwo = 0L
                var result: Long = minusOne
                for (i in 2..n) {
                    result = minusOne + minusTwo
                    minusTwo = minusOne
                    minusOne = result
                }
                return result
            }
        }
    }

    private fun throwNegativeArgumentException(): Nothing {
        throw IllegalArgumentException("n value cannot be negative number")
    }
}

File: iOS/FrontendTechnicalChallenge/FrontendTechnicalChallenge/Fibonacci.swift
//
//  Fibonacci.swift
//  FrontendTechnicalChallenge
//
//  Created by Ahmad Mahmoud on 11/12/2020.
//

import Foundation

struct Fibonacci {
    
    static func recursive(n: Int64) throws -> Int64 {
        if(n < 0) {
            throw Errors.IllegalArgumentException(message: "n value cannot be negative number")
        }
        else if (n < 2) {
            return n
        }
        else {
            return try! recursive(n: n - 1) + recursive(n: n - 2)
        }
    }
    
    static func iterative(n: Int64) throws -> Int64 {
        if(n < 0) {
            throw Errors.IllegalArgumentException(message: "n value cannot be negative number")
        }
        else if (n < 2) {
            return n
        }
        else {
            var minusOne: Int64 = 1
            var minusTwo: Int64 = 0
            var result: Int64 = Int64(minusOne)
            for _ in 2...n {
                result = minusOne + minusTwo
                minusTwo = minusOne
                minusOne = result
            }
            return result
        }
    }
}
